api/client/credentials.go

```diff
@@ -0,0 +1,154 @@
+limitations under the License.
+*/
+
+package client
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"io/ioutil"
+
+	"github.com/gravitational/teleport/api/constants"
+	"github.com/gravitational/trace"
+	"golang.org/x/net/http2"
+)
+
+// Credentials are used to authenticate the client's connection to the server.
+type Credentials struct {
+	TLS *tls.Config
+}
+
+// CheckAndSetDefaults checks and sets default credential values.
+func (c *Credentials) CheckAndSetDefaults() error {
+	if c.TLS == nil {
+		return trace.BadParameter("missing TLS config")
+	}
+	c.TLS = c.TLS.Clone()
+	c.TLS.NextProtos = []string{http2.NextProtoTLS}
+	if c.TLS.ServerName == "" {
+		c.TLS.ServerName = constants.APIDomain
+	}
+	return nil
+}
+
+// LoadIdentityFile attempts to load credentials from the specified identity file path.
+// An identity file can be saved to disk by running `tsh login --out=identity_file_path`.
+func LoadIdentityFile(path string) (*Credentials, error) {
+	idf, err := ReadIdentityFile(path)
+	if err != nil {
+		return nil, trace.BadParameter("identity file could not be decoded: %v", err)
+	}
+
+	tls, err := idf.TLS()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	creds, err := LoadTLS(tls)
+	return creds, trace.Wrap(err)
+}
+
+// LoadKeyPair attempts to load credentials from a certificate key pair and root CAs.
+// Those files can be saved to disk with `tctl auth sign --out=path`.
+// EX: path=/certs/admin creates three files - /certs/admin.(key|crt|cas).
+func LoadKeyPair(crtFile, keyFile, casFile string) (*Credentials, error) {
+	cert, err := tls.LoadX509KeyPair(crtFile, keyFile)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	caCerts, err := ioutil.ReadFile(casFile)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	pool := x509.NewCertPool()
+	if ok := pool.AppendCertsFromPEM(caCerts); !ok {
+		return nil, trace.Errorf("invalid TLS CA cert PEM")
+	}
+
+	creds, err := LoadTLS(&tls.Config{
+		Certificates: []tls.Certificate{cert},
+		RootCAs:      pool,
+	})
+	return creds, trace.Wrap(err)
+}
+
+// LoadTLS returns Credentials with the given TLS config.
+func LoadTLS(tls *tls.Config) (*Credentials, error) {
+	creds := &Credentials{TLS: tls}
+	if err := creds.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return creds, nil
+}
+
+// CredentialsProvider has a defined source of credentials, which can be dynamically loaded.
+// CredentialsProviders are used to load and test credentials during client initialization.
+// Use the function NewIdentityFileProvider or NewKeyPairProvider to create a CredentialLoader.
+type CredentialsProvider interface {
+	// Load attempts to load credentials from the provider.
+	Load() (*Credentials, error)
+}
+
+// NewIdentityFileProvider returns a CredentialsProvider that uses an identity file to load credentials.
+// An identity file can be saved to disk by running `tsh login --out=path`.
+func NewIdentityFileProvider(path string) CredentialsProvider {
```

Comment:
If you need to test this, returning a concrete type is better than an interface if you need APIs specific to the tested implementation. Instead, use an exported type (e.g. `IdentityFileProvider`) but with all attributes unexported.